pkg/client/intermediate/intermediate.go

```diff
@@ -0,0 +1,139 @@
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"sort"
+
+	"github.com/bitnami-labs/charts-syncer/pkg/client"
+
+	"github.com/juju/errors"
+
+	"github.com/bitnami-labs/charts-syncer/internal/utils"
+	"github.com/bitnami-labs/charts-syncer/pkg/client/types"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+var (
+	versionRe = regexp.MustCompile("(.*)-(\\d+\\.\\d+\\.\\d+)\\.bundle.tar")
+)
+
+// Repo allows to operate a chart repository.
+type Repo struct {
+	dir     string
+	entries map[string][]string
+}
+
+// NewIntermediateClient returns a ReadWriter object
+var NewIntermediateClient = func(intermediateBundlesPath string) (client.ReadWriter, error) {
+	return New(intermediateBundlesPath)
+}
+
+// New creates a Repo object from an api.Repo object.
+func New(dir string) (*Repo, error) {
+	d, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := os.MkdirAll(d, 0755); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Populate entries from directory
+	entries := make(map[string][]string)
+	matches, err := filepath.Glob(filepath.Join(d, "*.tar"))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, m := range matches {
+		filename := filepath.Base(m)
+		s := versionRe.FindStringSubmatch(filename)
+		entries[s[1]] = append(entries[s[1]], s[2])
+		sort.Strings(entries[s[0]])
+	}
+
+	return &Repo{dir: d, entries: entries}, nil
+}
+
+// List lists all chart names in a repo
+func (r *Repo) List() ([]string, error) {
+	var names []string
+	for name := range r.entries {
+		names = append(names, name)
+	}
+	return names, nil
+}
+
+// ListChartVersions lists all versions of a chart
+func (r *Repo) ListChartVersions(name string) ([]string, error) {
+	versions, ok := r.entries[name]
+	if !ok {
+		return []string{}, nil
+	}
+	return versions, nil
+}
+
+// Fetch fetches a chart
+func (r *Repo) Fetch(name string, version string) (string, error) {
+	return path.Join(r.dir, fmt.Sprintf("%s-%s.bundle.tar", name, version)), nil
+}
+
+// Has checks if a repo has a specific chart
+func (r *Repo) Has(name string, version string) (bool, error) {
+	versions, err := r.ListChartVersions(name)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	for _, v := range versions {
+		if v == version {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// Upload uploads a chart to the repo
+func (r *Repo) Upload(filepath string, metadata *chart.Metadata) error {
+	name := metadata.Name
+	version := metadata.Version
+	if _, ok := r.entries[name]; ok {
+		for _, v := range r.entries[name] {
+			if v == version {
+				return errors.AlreadyExistsf("%s-%s", name, version)
+			}
+		}
+	}
+
+	input, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return errors.Annotatef(err, "reading %q", filepath)
+	}
+
+	out := path.Join(r.dir, fmt.Sprintf("%s-%s.bundle.tar", name, version))
+	if err := ioutil.WriteFile(out, input, 0644); err != nil {
+		return errors.Annotatef(err, "creating %q", out)
+	}
```

Comment:
How big is what is being read at `filepath`?
I would rather copy streams...
```suggestion
	src, err := os.Open(filepath)
	if err != nil {
		return errors.Annotatef(err, "reading %q", filepath)
	}

	out := path.Join(r.dir, fmt.Sprintf("%s-%s.bundle.tar", name, version))
	dst, err = os.Create(out, 0644)
	if err != nil {
		return errors.Annotatef(err, "creating %q", out)
	}
	_, err = io.Copy(dst, src)
	if err != nil {
		return errors.Annotatef(err, "copying %q to %q", filepath, out)
	}
```